File: 3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.py
class Solution:
    def earliestFinishTime(self, ls: List[int], ld: List[int], ws: List[int], wd: List[int]) -> int:
        n=len(ls)
        m=len(ws)

        first=float('inf')
        for i in range(n):
            first=min(first, ls[i]+ld[i])

        res=float('inf')
        for i in range(m):
            res=min(res, wd[i]+max(first, ws[i]))

        first=float('inf')
        for i in range(m):
            first=min(first, ws[i]+wd[i])
        for i in range(n):
            res=min(res, ld[i]+max(first, ls[i]))

        return res
        

File: Easy/3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.java
class Solution{
    public int earliestFinishTime(int[] ls, int[] ld, int[] ws, int[] wd){
        int n=ls.length;
        int m=ws.length;

        int first=Integer.MAX_VALUE;
        for(int i=0; i<n; i++){
            first=Math.min(first, ls[i]+ld[i]);
        }

        int res=Integer.MAX_VALUE;
        for(int i=0; i<m; i++){
            res=Math.min(res, wd[i]+Math.max(first, ws[i]));
        }

        first=Integer.MAX_VALUE;
        for(int i=0; i<m; i++){
            first=Math.min(first, ws[i]+wd[i]);
        }
        for(int i=0; i<n; i++){
            res=Math.min(res, ld[i]+Math.max(first, ls[i]));
        }

        return res;

    }
}